### PR TITLE
UAVCAN F9P Moving Baseline

### PR DESCRIPTION
**Pull request by AlexKlimaj**
This PR adds moving baseline functionality on the ARK RTK GPS using the U-Blox F9P.

Requires https://github.com/PX4/PX4-GPSDrivers/pull/91

PXL_20211011_203619308

### CODE REVIEW

**Comment by AlexKlimaj** [comment]
I got some test flying in today but it looks like the logs are mostly corrupt.
[22_01_56.zip](https://github.com/PX4/PX4-Autopilot/files/7356376/22_01_56.zip)

**Comment by AlexKlimaj** [comment]
Adding setting CAN Priorities in this PR. Reduces error rate drastically. Without priorities set explicitly, they are all the same and collisions can't resolve until they both stop trying.

RTCM data is the largest amount of data and there is a tight timing requirement to get from one RTK GPS to the other, so I set that to max. Then GPS data one lower than that. After that, regular sensors are default, and messages such as Battery Status are set to low.

Current with all default priorities.
CAN No Priorities Set High Errors

With priorities. Two RTK GPS cannodes and a CAN battery.
CAN Priorities Set No Errors

With priorities. Two RTK GPS cannodes, CAN battery, ARK CAN GPS, and two ARK Flows. A few errors start showing up. Likely from having many GPS, Mag, Baro, and Flow messages on the bus.
CAN Priorities Set High Load

**Comment by AlexKlimaj** [comment]
Update to point at https://github.com/PX4/PX4-GPSDrivers/pull/92 commit when it is merged.

File: src/drivers/uavcannode/Subscribers/MovingBaselineData.hpp [hunk 1]
```diff
@@ -72,23 +72,34 @@ class RTCMStream :
...
-		memcpy(gps_inject_data.data, &msg.data[0], gps_inject_data.len);
+			gps_inject_data.len = msg.data.size();
 
-		gps_inject_data.timestamp = hrt_absolute_time();
-		_gps_inject_data_pub.publish(gps_inject_data);
+			memcpy(gps_inject_data.data, &msg.data[0], gps_inject_data.len);
+
+			gps_inject_data.timestamp = hrt_absolute_time();
+
+			gps_inject_data.device_id = msg.getSrcNodeID().get();
```
Comment (dagar) [inline src/drivers/uavcannode/Subscribers/MovingBaselineData.hpp:99]: A valid device id would have the bus type set to UAVCAN, the bus, and the node id as the address. https://github.com/PX4/PX4-Autopilot/blob/fb54324f5612d5565cfbf366d295799a0641dd22/src/lib/drivers/device/Device.hpp#L160-L165

For the device type you can use the generic DRV_GPS_DEVTYPE_UAVCAN as unfortunately we don't know the specific type over CAN.

https://github.com/PX4/PX4-Autopilot/blob/fb54324f5612d5565cfbf366d295799a0641dd22/src/drivers/drv_sensor.h#L161

**Review by dagar — DISMISSED** [review]

**Comment by dagar** [comment]
This looks good from my perspective. As for the memory usage of `gps_inject_data` could we consider cutting the queue depth in half (8->4) the next time it's tested on a "bad" setup (wifi or other lossy link)?

**Comment by bkueng** [comment]
> This looks good from my perspective. As for the memory usage of gps_inject_data could we consider cutting the queue depth in half (8->4) the next time it's tested on a "bad" setup (wifi or other lossy link)?

Can you instead keep the message size and accumulate multiple topics? The current size fits the mavlink message, and increasing it wastes that (almost 1KB now), independent from queue size.

**Comment by dagar** [comment]
As discussed let's get this in and consider queue or buffer adjustments in a followup.

File: src/drivers/uavcannode/Publishers/GnssFix2.hpp [hunk 2]
```diff
@@ -118,6 +120,26 @@ class GnssFix2 :
 			fix2.covariance.push_back(gps.s_variance_m_s);
 			fix2.covariance.push_back(gps.s_variance_m_s);
 
+			uavcan::equipment::gnss::ECEFPositionVelocity ecefpositionvelocity{};
+			ecefpositionvelocity.velocity_xyz[0] = NAN;
+			ecefpositionvelocity.velocity_xyz[1] = NAN;
+			ecefpositionvelocity.velocity_xyz[2] = NAN;
+
+			// Use ecef_position_velocity for now... There is no heading field
```
Comment (benflyvoly) [inline src/drivers/uavcannode/Publishers/GnssFix2.hpp:128]: This makes absolutely no sense. Worse than confusing track angle and heading. Only good news is that we don't use ECEF in our current configuration. We should use it for long missions and earth rotation sensitive sensors in the future.

**Pull request merged**